src/musig.rs

```diff
@@ -0,0 +1,1158 @@
+//! This module implements high-level Rust bindings for a Schnorr-based
+//! multi-signature scheme called MuSig2 [paper](https://eprint.iacr.org/2020/1261).
+//! It is compatible with bip-schnorr.
+//!
+//! The documentation in this module is for reference and may not be sufficient
+//! for advanced use-cases. A full description of the C API usage along with security considerations
+//! can be found in [C-musig.md](secp256k1-sys/depend/secp256k1/src/modules/musig/musig.md).
+use core;
+use core::fmt;
+use core::mem::MaybeUninit;
+#[cfg(feature = "std")]
+use std;
+
+use crate::ffi::{self, CPtr};
+use crate::{
+    schnorr, Error, Keypair, Message, PublicKey, Scalar, Secp256k1, SecretKey, Signing, Verification, XOnlyPublicKey
+};
+
+/// Musig partial signature parsing errors
+#[derive(Debug, Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub enum ParseError {
+    /// Parse Argument is malformed. This might occur if the point is on the secp order,
+    /// or if the secp scalar is outside of group order
+    MalformedArg,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for ParseError {}
+
+impl fmt::Display for ParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        match *self {
+            ParseError::MalformedArg => write!(f, "Malformed parse argument"),
+        }
+    }
+}
+
+/// Session Id for a MuSig session.
+#[allow(missing_copy_implementations)]
+#[derive(Debug, Eq, PartialEq)]
+pub struct SessionSecretRand([u8; 32]);
+
+impl SessionSecretRand {
+    /// Generates a new session ID using thread RNG.
+    #[cfg(all(feature = "rand", feature = "std"))]
+    pub fn new() -> Self {
+        Self::from_rng(&mut rand::thread_rng())
+    }
+
+    /// Creates a new [`SessionSecretRand`] with random bytes from the given rng
+    #[cfg(feature = "rand")]
+    pub fn from_rng<R: rand::Rng + ?Sized>(rng: &mut R) -> Self {
+        let session_secrand = crate::random_32_bytes(rng);
+        SessionSecretRand(session_secrand)
+    }
+
+    /// Creates a new [`SessionSecretRand`] with the given bytes.
+    ///
+    /// Special care must be taken that the bytes are unique for each call to
+    /// [`KeyAggCache::nonce_gen`] or [`new_nonce_pair`]. The simplest
+    /// recommendation is to use a cryptographicaly random 32-byte value.
+    ///
+    /// In rand-std environment, [`SessionSecretRand::new`] can be used to generate a random
+    /// session id using thread rng.
+    pub fn assume_unique_per_nonce_gen(inner: [u8; 32]) -> Self { SessionSecretRand(inner) }
```

Comment:
In dd26c3067044a89c503a32079c24227d6e1afca5:

I think you should panic here if the user provides all 0s. Will probably be easier to diagnose than them getting a panic later when signing. (I wouldn't bother returning a Result or something ... this function is supposed to be passed the result of a cryptographic hash or RNG or something.) (Will do in followup.)

Reply:
Probably OK to do it later but this kind of thing tends to come up in cryptography more often so maybe we should have a newtype for it.

Reply:
Hard to say. This function is pretty-explicitly "I got these bytes from somewhere and they're legit, promise" so it feels both paternalistic and insufficient to provide a newtype that does a single validity check on it.

Since the upstream library will do bad things if you provide all zeros, I think we should catch that, but I think that immediately panicking is probably the least bad thing to do.

Reply:
I think in this case we should just keep it as-is. We sorta tried this with the `ThirtyTwoByteHash` trait and it wound up just causing a lot of unnecessary and frustrating coupling between libraries, and also the "per-nonce" uniqueness constraint here is stronger than what a rng library can necessarily produce.